Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp

```diff
@@ -434,32 +435,51 @@ RegisterID* ArrayNode::emitBytecode(BytecodeGenerator& generator, RegisterID* ds
             JSValue constant = static_cast<ConstantNode*>(firstPutElement->value())->jsValue(generator);
             if (UNLIKELY(!constant))
                 hadVariableExpression = true;
-            else
+            else {
                 recommendedIndexingType = leastUpperBoundOfIndexingTypeAndValue(recommendedIndexingType, constant);
+                if (!constant.isString())
+                    allDenseStrings = false;
+                else if (auto* impl = asString(constant)->tryGetValueImpl(); !impl || !impl->isAtom())
+                    allDenseStrings = false;
+            }
         }
 
         ++length;
     }
+    if (hadVariableExpression)
+        allDenseStrings = false;
 
-    auto newArray = [&] (RegisterID* dst, ElementNode* elements, unsigned length, bool hadVariableExpression) {
+    auto newArray = [&](RegisterID* dst, ElementNode* elements, unsigned length) {
         if (length && !hadVariableExpression) {
+            VM& vm = generator.vm();
             recommendedIndexingType |= CopyOnWrite;
-            ASSERT(generator.vm().heap.isDeferred()); // We run bytecode generator under a DeferGC. If we stopped doing that, we'd need to put a DeferGC here as we filled in these slots.
-            auto* array = JSImmutableButterfly::create(generator.vm(), recommendedIndexingType, length);
+            ASSERT(vm.heap.isDeferred()); // We run bytecode generator under a DeferGC. If we stopped doing that, we'd need to put a DeferGC here as we filled in these slots.
+
+            Structure* immutableButterflyStructure = allDenseStrings ? vm.immutableButterflyAtomStringsStructure.get() : vm.immutableButterflyStructure(recommendedIndexingType);
+            auto* array = JSImmutableButterfly::tryCreate(generator.vm(), immutableButterflyStructure, length);
+            RELEASE_ASSERT(array);
+
             unsigned index = 0;
             for (ElementNode* element = elements; index < length; element = element->next()) {
                 ASSERT(element->value()->isConstant());
                 JSValue constant = static_cast<ConstantNode*>(element->value())->jsValue(generator);
                 ASSERT(constant);
                 array->setIndex(generator.vm(), index++, constant);
+                if (allDenseStrings) {
+                    ASSERT(constant.isString());
+                    JSString* string = asString(constant);
+                    vm.atomStringToJSStringMap.set(const_cast<StringImpl*>(string->tryGetValueImpl()), string);
```

Comment:
Let's always register any JSString generated during BytecodeGenerator.
This does not ensure that `constant` is that string.

Reply:
It seems the AtomString pointer comparison has slightly better performance than JSString pointer comparison. This patch is based on the former one. So, in this case, I think we don't need to register all JSStrings since we use AtomString pointers for the comparisons. However, we do need to use the registered the JSString to update the butterfly.